plugins/generator-1.17.1/forge-1.17.1/procedures/spawn_particle.java.ftl

```diff
@@ -1 +1 @@
-world.addParticle(${generator.map(field$particle, "particles")}, ${input$x}, ${input$y}, ${input$z}, ${input$xs}, ${input$ys}, ${input$zs});
\ No newline at end of file
+world.addParticle((SimpleParticleType) ${generator.map(field$particle, "particles")}, ${input$x}, ${input$y}, ${input$z}, ${input$xs}, ${input$ys}, ${input$zs});
```

Comment:
Why was this cast added?

Reply:
`ParticleType<?>` doesn't implement `ParticleOptions`, so without it the build fails.

Reply:
Which particle uses this? 

If you just force cast it without checking, it could crash later in the game

Reply:
I think instead of this, use SimpleParticleType as return type in your register function and use SimpleParticleType in the registry fields

Reply:
<details><summary>Converted everything to `SimpleParticleType`</summary><p>Executing Gradle task: build
Build info: MCreator 2021.3.99999, forge-1.17.1, 64-bit, 8065 MB, Windows 7, JVM 16.0.1, JAVA_HOME: C:\Users\user\AppData\Roaming\Yandex\New Folder\New Folder\.minecraft\mcreation\text\other\external\Git\MCreator\jdk\jdk16_win_64
 
> Configure project :
The code of this workspace uses official obfuscation mappings provided by Mojang. These mappings fall under their associated license you should be fully aware of.
(c) 2020 Microsoft Corporation. These mappings are provided "as-is" and you bear the risk of using them. You may copy and use the mappings for development purposes,
but you may not redistribute the mappings complete and unmodified. Microsoft makes no warranties, express or implied, with respect to the mappings provided here.
Use and modification of this document or the source code (in any form) of Minecraft: Java Edition is governed by the Minecraft End User License Agreement available
at https://account.mojang.com/documents/minecraft_eula.
 
> Task :compileJava
C:\Users\user\AppData\Roaming\Yandex\New Folder\New Folder\.minecraft\mcreation\text\MCreatorWorkspaces\new_twentyfirst_three_copy\src\main\java\net\mcreator\newtwentyfirstthree\init\NewTwentyfirstThreeModParticles.java:33: error: incompatible types: ParticleType cannot be converted to SimpleParticleType  new SimpleParticleType(false).setRegistryName("exampleparticle_1"),
        ^
 where CAP#1 is a fresh type-variable:
 CAP#1 extends ParticleOptions from capture of ?
Note: Some messages have been simplified; recompile with -Xdiags:verbose to get full output
1 error
> Task :compileJava FAILED
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':compileJava'.
> Compilation failed; see the compiler error output for details.
* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
* Get more help at https://help.gradle.org
BUILD FAILED in 18s
1 actionable task: 1 executed
 
BUILD FAILED
Task completed in 21 seconds</p></details>

Reply:
<details><summary>Converted everything to `SimpleParticleType`</summary><p>Executing Gradle task: build
Build info: MCreator 2021.3.99999, forge-1.17.1, 64-bit, 8065 MB, Windows 7, JVM 16.0.1, JAVA_HOME: ...\MCreator\jdk\jdk16_win_64
 
> Configure project :
The code of this workspace uses official obfuscation mappings provided by Mojang. These mappings fall under their associated license you should be fully aware of.
(c) 2020 Microsoft Corporation. These mappings are provided "as-is" and you bear the risk of using them. You may copy and use the mappings for development purposes,
but you may not redistribute the mappings complete and unmodified. Microsoft makes no warranties, express or implied, with respect to the mappings provided here.
Use and modification of this document or the source code (in any form) of Minecraft: Java Edition is governed by the Minecraft End User License Agreement available
at https://account.mojang.com/documents/minecraft_eula.
 
> Task :compileJava
...\new_twentyfirst_three_copy\src\main\java\net\mcreator\newtwentyfirstthree\init\NewTwentyfirstThreeModParticles.java:33: error: incompatible types: ParticleType cannot be converted to SimpleParticleType  new SimpleParticleType(false).setRegistryName("exampleparticle_1"),
        ^
 where CAP#1 is a fresh type-variable:
 CAP#1 extends ParticleOptions from capture of ?
Note: Some messages have been simplified; recompile with -Xdiags:verbose to get full output
1 error
> Task :compileJava FAILED
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':compileJava'.
> Compilation failed; see the compiler error output for details.
* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
* Get more help at https://help.gradle.org
BUILD FAILED in 18s
1 actionable task: 1 executed
 
BUILD FAILED
Task completed in 21 seconds</p></details>

Reply:
>If you just force cast it without checking, it could crash later in the game

Currently we can only use particles in procedures and
>particle command

that seem to work, so we'll likely have to wait reporting those problems when porting block/entity/item.

Reply:
>I think instead of this, use SimpleParticleType as return type in your register function and use SimpleParticleType in the registry fields

I used ParticleType<?> throughout the registry 'cause it's a common class for all particle types, including SimpleParticleType.

Reply:
But there is no need to, you can use SimpleParticleType as all custom particles are of this type and remove the redundant cast then.